### ISSUE TITLE
Shared class issue (definition not cleared on plugin reload) related to exceptions

### ISSUE THREAD

**Issue by XertroV** [openplanet-nl/openplanet-issues#451]
I've been having a few issues developing some shared class stuff recently where the definitions don't update till I restart the game (and sometimes I will get errors about mismatching shared classes, but sometimes it compiles fine just not updated).

This might be related to exceptions, especially exceptions during hooks (but I am not sure if this is relevant).

For example, changing the following code didn't update when reloading the plugin (I forgot to add @ to the beginning of this.block = block and this.item = item). After reloading, I got the same exceptions and no compile errors (the exception was null ptr exception b/c it was probably looking for .opAssign):

```angelscript
    shared class SetSkinSpec : NetworkSerializable {
        string fgSkin;
        string bgSkin;
        BlockSpec@ block;
        ItemSpec@ item;

        SetSkinSpec(BlockSpec@ block, const string &in fgSkin, const string &in bgSkin) {
            this.block = block;
            this.fgSkin = fgSkin;
            this.bgSkin = bgSkin;
        }

        SetSkinSpec(ItemSpec@ item, const string &in skin, bool isForegroundElseBackground) {
            this.item = item;
            if (isForegroundElseBackground) {
                this.fgSkin = skin;
            } else {
                this.bgSkin = skin;
            }
        }
    }
```

**Comment by codecat** [comment]
I am unable to reproduce this 🧐

Is the shared class used by any other dependant plugins? I just tested it with `main()` copy to a separate dependant plugin and it still works for me.

My test code:
```asc
shared class NetworkSerializable{}
shared class BlockSpec{}
shared class ItemSpec{}

shared class SetSkinSpec : NetworkSerializable {
	string fgSkin;
	string bgSkin;
	BlockSpec@ block;
	ItemSpec@ item;

	SetSkinSpec(BlockSpec@ block, const string &in fgSkin, const string &in bgSkin) {
		@this.block = block;
		this.fgSkin = fgSkin;
		this.bgSkin = bgSkin;
	}

	SetSkinSpec(ItemSpec@ item, const string &in skin, bool isForegroundElseBackground) {
		@this.item = item;
		if (isForegroundElseBackground) {
			this.fgSkin = skin;
		} else {
			this.bgSkin = skin;
		}
	}
}
```
```asc
void Main()
{
	SetSkinSpec test(BlockSpec(), "", "");
	SetSkinSpec test2(ItemSpec(), "", false);
}
```

**Comment by XertroV** [comment]
> Is the shared class used by any other dependant plugins? 

Yes, and I reload the dependency plugin (which should reload both)

I think changing a function body but not signature will not trigger an error but the old function logic applies. 
Changing function signatures does produce an error (these are function signatures of methods of the shared class)

**Comment by XertroV** [comment]
This error can be triggered by changing to school/dev mode **even though the file is exactly the same**. I have no idea how this would be possible.

The file in quesiton has no preprocessor statements or really anything funky going on. it's just some arrays and bools. 

https://github.com/XertroV/tm-editor-plus-plus/blob/master/src/Editor/DrawLinesAndQuads_Shared.as

Also to be clear, there has only ever been 1 published version of this file.

![image](https://github.com/user-attachments/assets/b0e54c45-ab36-4fcc-bab5-711c73319bf5)

**Comment by XertroV** [comment]
okay managed to repro. It requires a dependent plugin. I used optional dependency b/c that's what most things use for E++. 

screenshot shows repro after reload. 

repro should be: 
- Load plugins
- uncomment defines in info.toml of repro-shared-code-issue-dev 
- reload plugins

![image](https://github.com/user-attachments/assets/a06f7591-ee7a-41c1-9aa0-27e4279faa76)
[repro-shared-issue.zip](https://github.com/user-attachments/files/17256248/repro-shared-issue.zip)

**Comment by codecat** [comment]
Thanks! I will check this 👍

**Comment by XertroV** [comment]
Any updates on this?

Edit: also got this today by changing an `array<int2>` to an `array<int>` in mlfeed.

**Comment by codecat** [comment]
I suspect what you may be running into is related to your `info.toml` defines. Those are per-plugin, and will currently not be carried over to its dependencies. So, in this case, you will have unmatching declarations because the other plugin did not use the `ENABLE_PROBLEM` define in its shared code compilation.

That means that currently this is by design. Whether they should be carried over from dependencies is up for debate. It's not a very straightforward fix.

**Comment by XertroV** [comment]
The info toml stuff is just for demonstrating it. You can manually save the files in the same way, I just wanted to make it easier for you.
(When this bug happens to me, I don't change anyhting in any info.tomls)

**Comment by codecat** [comment]
In that case I am still unable to reproduce it - editing the files manually made no difference and it worked as expected.

**Comment by XertroV** [comment]
Hmm okay yeah I can't reproduce it now either. 

Definitely still happens for both MLFeed and Editor++ tho. :/

Both use defines but they don't change and I don't think I use any `#if DEV` type things in the shared exports.

**Comment by codecat** [comment]
Does this still happen frequently?

**Comment by XertroV** [comment]
Yes. haven't tried the conditions on edge version yet (update: didn't help)

**Comment by XertroV** [comment]
Okay I got this problem with this fairly simple class but not a minimal reproduction.

```asc
namespace VoiceLines {
    shared class IVoiceLineParams {
        bool isUrl;
        string pathOrUrl;
        string subtitles;
        string imagePathOrUrl;
        IVoiceLineParams(const string &in pathOrUrl, const string &in subtitles, const string &in imagePathOrUrl = "") {
            isUrl = pathOrUrl.StartsWith("https://") || pathOrUrl.StartsWith("http://");
            this.pathOrUrl = pathOrUrl;
            this.subtitles = subtitles;
            this.imagePathOrUrl = imagePathOrUrl;
        }
    }
}
```

Rough setup:
- I had dips++ and dips++-editor running with a dependency setup and some function exports.
- I added `namespace VoiceLines { shared class IVoiceLineParams {} }` and reloaded dips++ which also reloaded dips++-editor.
- I added the properties and constructor.
- I reloaded dips++ again and got the "doesn't match the original declaration in other module" error.

**Comment by codecat** [comment]
Was unable to reproduce with the above code sample and several different attempts.

If this happens again, can you shut down the game and post a full log file please?

**Comment by codecat** [comment]
I opened a potential relevant issue on the Angelscript forum: https://www.gamedev.net/forums/topic/718724-shared-types-refcounting-bug/

Here are 3 plugins (TestWM, UME, WM) that reproduce the issue in the forum thread. It's not a reproduction of the "doesn't match original declaration" error, but there's a good chance they are related.

[Repro.zip](https://github.com/user-attachments/files/20967789/Repro.zip)
